### PR TITLE
Support sharding subqueries in instant query QF

### PR DESCRIPTION
**Pull request by yeya24**
**What this PR does**:

https://github.com/thanos-io/thanos/issues/5811

Thanos doesn't support query sharding for subqueries.
This PR updates the Thanos version to get it supported and also support Matrix response type in instant query so that we can merge the sharded sub query responses.

**Which issue(s) this PR fixes**:
Fixes #<issue number>

**Checklist**
- [x] Tests updated
- [ ] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Comment by yeya24** [comment]
cc @alanprot @alvinlin123

File: pkg/querier/tripperware/utils/utils.go [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+import (
+	"sort"
+
+	"github.com/cortexproject/cortex/pkg/cortexpb"
+)
+
+// SliceSamples assumes given samples are sorted by timestamp in ascending order and
+// return a sub slice whose first element's is the smallest timestamp that is strictly
+// bigger than the given minTs. Empty slice is returned if minTs is bigger than all the
+// timestamps in samples.
```
Comment (alvinlin123) [inline pkg/querier/tripperware/utils/utils.go:12]: nit: I think it's always a good idea to explicit called out what happens if caller violates assumption. Like "If the given samples is not sorted, the behaviour is undefined". Of course it's debatable that this is just stating the obvious; hence the nit :)

File: pkg/querier/tripperware/instantquery/instant_query.go [hunk 2]
```diff
@@ -324,6 +335,59 @@ func vectorMerge(resps []*PrometheusInstantQueryResponse) *Vector {
 	return result
 }
 
+func matrixMerge(resps []*PrometheusInstantQueryResponse) *Matrix {
```
Comment (alvinlin123) [inline pkg/querier/tripperware/instantquery/instant_query.go:338]: This function looks a lot like https://github.com/cortexproject/cortex/blob/997e4af542e8363b02ddb759bad4019cef9d1692/pkg/querier/tripperware/queryrange/query_range.go#L347 

Anyway we can reduce the duplication?

File: pkg/storegateway/bucket_stores.go [hunk 3]
```diff
@@ -487,6 +487,7 @@ func (u *BucketStores) getOrCreateStore(userID string) (*store.BucketStore, erro
 		u.syncDirForUser(userID),
 		newChunksLimiterFactory(u.limits, userID),
 		newSeriesLimiterFactory(u.limits, userID),
+		store.NewBytesLimiterFactory(0),
```
Comment (alvinlin123) [inline pkg/storegateway/bucket_stores.go:490]: I assume this new parameter is due to new Thanos dependency. Does this parameter means "unlimited" and is the "old behaviour"? Mainly asking for confirmation :)

Comment (yeya24) [reply pkg/storegateway/bucket_stores.go:490]: Unlimited

File: pkg/querier/tripperware/instantquery/instant_query.go (hunk 2 above)
Comment (yeya24) [reply pkg/querier/tripperware/instantquery/instant_query.go:338]: No good way for now cause they are different types.

Comment (yeya24) [reply pkg/querier/tripperware/instantquery/instant_query.go:338]: We could expose an interface for sure... But this needs more refactoring. I prefer copying the function for now.

File: pkg/querier/tripperware/utils/utils.go (hunk 1 above)
Comment (yeya24) [reply pkg/querier/tripperware/utils/utils.go]: Updated

File: pkg/querier/tripperware/utils/utils.go [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+func SliceSamples(samples []cortexpb.Sample, minTs int64) []cortexpb.Sample {
```
[the remark this answers is not in the document]
Comment (yeya24) [reply pkg/querier/tripperware/utils/utils.go]: Cool idea. I added an unit test

File: pkg/querier/tripperware/instantquery/instant_query.go (hunk 2 above)
Comment (alvinlin123) [reply pkg/querier/tripperware/instantquery/instant_query.go:337]: From my brief look, I think we can break https://github.com/cortexproject/cortex/blob/997e4af542e8363b02ddb759bad4019cef9d1692/pkg/querier/tripperware/queryrange/query_range.go#L347 into multiple functions one that takes in `[]tripperware.SampleStream` and the other that takes in `resps []*PrometheusResponse`. 

Then here you may be able to use the new function that takes in `[]tripperware.SampleStream`.

Comment (yeya24) [reply pkg/querier/tripperware/instantquery/instant_query.go]: Updated. Could you please take another look?

**Review by alvinlin123 — APPROVED** [review]

**Review by alanprot — APPROVED** [review]

**Review by alanprot — APPROVED** [review]

**Pull request merged**